init/devices.cpp

```diff
@@ -399,44 +411,112 @@ std::vector<std::string> DeviceHandler::GetBlockDeviceSymlinks(const Uevent& uev
     return links;
 }
 
-void DeviceHandler::HandleDevice(const std::string& action, const std::string& devpath,
+void DeviceHandler::HandleDevice(const std::string& action, const std::string& dev_path,
                                  const std::string& upath, bool block,
                                  int major, int minor, const std::vector<std::string>& links) const {
-    std::vector<std::string> all_links(links);
-    std::string alias_link;
+    std::string state_link_path, alias_link_path;
+    std::stringstream slfmt;
 
-    // Add the alias link to the device to the list of links to
-    // create.
-    if (GetDeviceAlias(upath, major, minor, alias_link))
-        all_links.push_back(alias_link);
+    // Create a symlink in /dev/aliases that will point to the
+    // alias.
+    slfmt << "/dev/aliases/" << major << "_" << minor;
+    state_link_path = slfmt.str();
 
     if (action == "add") {
-        MakeDevice(devpath, block, major, minor, all_links);
-        for (const auto& link : all_links) {
-            if (!mkdir_recursive(Dirname(link), 0755)) {
-                PLOG(ERROR) << "Failed to create directory " << Dirname(link);
+        std::string alias_path;
+        std::vector<std::string> all_links(links);
+
+        // Add the alias link to the device to the list of links to
+        // create.
+        if (GetDeviceAlias(upath, major, minor, alias_link_path))
+            all_links.push_back(alias_link_path);
+
+        MakeDevice(dev_path, block, major, minor, all_links);
+
+        for (const auto& link_path : all_links) {
+            std::string link_target;
+
+            if (!mkdir_recursive(Dirname(link_path), 0755))
+                PLOG(ERROR) << "Failed to create directory " << Dirname(link_path);
+
+            // Only create a link in /dev/aliases for the device
+            // aliases we created from an 'alias' line in the
+            // configuration. This is a way to save what aliases we've
+            // created in /dev for a specific device and allow use to
+            // find the exact aliases we created without keeping state
+            // within ueventd.
+
+            // alias_link is the link (eg: /dev/modem -> /dev/ttyACM0)
+            // that we want to be creating.
+            if (!alias_link_path.empty()
+                && link_path == alias_link_path
+                && symlink(alias_link_path.c_str(), state_link_path.c_str())) {
+
+                // If creating the symlink failed, it can be because
+                // of a misc error, or it can be because the file
+                // already exists. We don't want to replace the file
+                // so we make it an explicit error message that the
+                // configuration is wrong.
+
+                if (errno != EEXIST)
+                    PLOG(ERROR) << "Failed to create alias symlink from " << state_link_path
+                                << " to " << alias_link_path;
```

Comment:
The log message in the `if` and `else` part of this code show the link name and its target in opposite order.

One reads "symlink _from_ linkpath to target" and the other "symlink target to linkpath". Which is confusing.